ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -62,6 +62,28 @@ private class ContractsFetch(
   }
 
   def contractsIo(jwt: Jwt, party: domain.Party, templateId: domain.TemplateId.RequiredPkg)(
+      implicit ec: ExecutionContext,
+      mat: Materializer): ConnectionIO[OffsetBookmark[domain.Offset]] = {
+
+    import doobie.postgres._
+    import doobie.implicits._
+
+    def loop(maxAttempts: Int): ConnectionIO[OffsetBookmark[domain.Offset]] = {
+      logger.debug(s"contractsIo, maxAttempts: $maxAttempts")
+      contractsIo_(jwt, party, templateId).exceptSql {
+        case e if maxAttempts > 0 && e.getSQLState == sqlstate.class23.UNIQUE_VIOLATION.value =>
```

Comment:
A uniqueness violation is not the only form of failure we should retry. For example, if two concurrent transactions attempt to change the offset (which was read at beginning of transaction) to different values, that should be retried as well.

Reply:
this is being caught by checking the number of rows updated by:
```
sql"""UPDATE ledger_offset SET last_offset = $newOffset where ... AND last_offset = $lastOffset"""
```